app/src/main/java/mozilla/lockbox/store/AutoLockStore.kt

```diff
@@ -0,0 +1,51 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.lockbox.store
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.os.Build
+import android.preference.PreferenceManager
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.rxkotlin.addTo
+import mozilla.lockbox.action.LifecycleAction
+import mozilla.lockbox.flux.Dispatcher
+import mozilla.lockbox.support.Constant
+import java.time.Instant
+
+class AutoLockStore(
+    val dispatcher: Dispatcher = Dispatcher.shared,
+    val settingStore: SettingStore = SettingStore.shared,
+    val lifecycleStore: LifecycleStore = LifecycleStore.shared
+) : ContextStore {
+
+    private val compositeDisposable = CompositeDisposable()
+    private lateinit var preferences: SharedPreferences
+
+    init {
+        lifecycleStore.lifecycleFilter
+            .filter { it == LifecycleAction.Background }
+            .switchMap { settingStore.autoLockTime.take(1) }
+            .subscribe {
+                val autoLockTimerDate = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    Instant.now().plusMillis(it.ms).toEpochMilli()
+                } else {
+                    TODO("VERSION.SDK_INT < O")
+                }
+
+                preferences
+                    .edit()
+                    .putLong(Constant.Key.autoLockTimerDate, autoLockTimerDate)
+                    .apply()
```

Comment:
Putting `autoLockTimerDate` and `autoLockTime` in `preferences` — rather than `SecurePreferences` — makes me a little nervous.

Reply:
@linuxwolf do you have an opinion about this?

Reply:
Throwing everything into `SecurePreferences` is an anti-pattern not to be repeated.

Things that go through `SecurePreferenecs` are ones where an observer can do something with them later -- such as the encryption keys or the user's information.  Settings like this do not have that concern, so shouldn't go into `SecurePreferences`.

Reply:
@linuxwolf I don't understand this reasoning.

I'm working on the assumption that `SharedPreferences` is not good a good place for secrets to be put — we wish to prevent unauthorized reads. And that `autolockDateTime` should not be written by anyone except us, lest it be written as an expired time — we wish to prevent unauthorized writes.

Reply:
in our implementation, `SecurePreferences` is just an encryption-layer wrapper for `SharedPreferences`, so we shouldn't have the same issues clearing it that we do on iOS.

Reply:
[ moved to keep the threading ]

> I'm working on the assumption that SharedPreferences is not good a good place for secrets to be put — we wish to prevent unauthorized reads. And that autolockDateTime should not be written by anyone except us, lest it be written as an expired time — we wish to prevent unauthorized writes.

As @sashei indicates, `SecurePreferenes` is a wrapper over `SharedPreferences`.  This means taht encrypting does not prevent unauthorized writes.  It means we might reject the value if it can't be decrypted, but then the item could be deleted instead and have a similar effect.  It's also not clear what exactly the attack this prevents and is worth trying to protect against.

Exfiltration of user information is a well-understood attack that we can definitely do something about.  The value of a user's personal (email, displayname, avatar, uid) and sensitive (encryption keys, credentials) information is high value; the value of `autolockTimerDate` and other settings is not.

Reply:
Here's my possible attack vector (@linuxwolf lmk what you think): someone could write a later datetime to the `autolocktime` shard preference if they knew how we were going to be reading it. We would then read the later date time as the next lock time, see that the system clock is behind that, and unlock. At least behind the encryption layer, the value will get decrypted to garbage and we can just lock when that happens.

Reply:
Ok, and the attacker does what with that?  What does the attack actually do?  Does it leak data?  Does it give the attack access to the database?

This is the problem I have with these.  I'm all for protecting data, but let's be smart about what we are protecting and why.  Protecting encryption keys and user info means an attacker can't then use that to get real data.